hack/kind-install-for-capk.sh

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env bash
+
+# Copyright 2025 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# This script installs a local Kind cluster with a local container registry and the correct files
+# mounted for using CAPK to test Cluster API.
+# The default Kind CNI is disabled because it doesn't work CAPK. Instead, Calico is used.
+# MetalLB is used as a local layer 2 load balancer to support exposing the API servers of workload
+# clusters on the local machine that's running this script.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+CALICO_VERSION=${CAPI_CALICO_VERSION:-"v3.29.1"}
+METALLB_VERSION=${CAPI_METALLB_VERSION:-""}
+# Set manually for non-Docker runtimes - example: "172.20"
+METALLB_IP_PREFIX=${CAPI_METALLB_IP_PREFIX:-""}
+KUBEVIRT_VERSION=${CAPI_KUBEVIRT_VERSION:-""}
+# Required to support pulling KubeVirt container disk images from private registries as well as
+# avoid Docker Hub rate limiting
+KIND_DOCKER_CONFIG_PATH=${CAPI_KIND_DOCKER_CONFIG_PATH:-"$HOME/.docker/config.json"}
+
+# Deploy Kind cluster
+KIND_CLUSTER_CONFIG="$(cat <<EOF
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+networking:
+  disableDefaultCNI: true
+nodes:
+- role: control-plane
+  extraMounts:
+  - containerPath: /var/lib/kubelet/config.json
+    hostPath: ${KIND_DOCKER_CONFIG_PATH}
+containerdConfigPatches:
+- |-
+  [plugins."io.containerd.grpc.v1.cri".registry]
+    config_path = "/etc/containerd/certs.d"
+EOF
+)"
+export KIND_CLUSTER_CONFIG
+
+"$(dirname "${BASH_SOURCE[0]}")/kind-install.sh"
+
+# Deploy Calico
+kubectl apply -f \
+  "https://raw.githubusercontent.com/projectcalico/calico/${CALICO_VERSION}/manifests/calico.yaml"
+
+# Deploy MetalLB
+if [[ -z "$METALLB_VERSION" ]]; then
+  METALLB_VERSION=$(curl "https://api.github.com/repos/metallb/metallb/releases/latest" \
+    | jq -r ".tag_name")
+fi
+
+kubectl apply -f \
+  "https://raw.githubusercontent.com/metallb/metallb/${METALLB_VERSION}/config/manifests/metallb-native.yaml"
+
+echo "Waiting for controller pod to be created..."
+while [[ -z $(kubectl -n metallb-system get pods \
+  -l app=metallb,component=controller -o jsonpath="{.items[0].metadata.name}" 2>/dev/null) ]]; do
+  sleep 2
+done
```

Comment:
It seems to waiting for MetalLB Deployment and DaemonSet pods can be a bit simplified with `kubectl wait` option. 

```
kubectl wait --for condition=available deployment controller  -n metallb-system --timeout 150s
kubectl wait --for condition=ready pod -l app=metallb,component=speaker -n metallb-system --timeout 150s
```

Reply:
I used the loop because I kept getting errors because the script was trying to check for `Ready` status on a non-existent pod.

Good point regarding checking for `available` status on the **deployment** rather than the pods. This works. But looks like we don't have an equivalent for daemon sets.

We could do what you suggested and check for `available` deployment and `ready` pod, however there will be a race between speaker pod creation and the wait on speaker pod.

Reply:
I've adopted your suggestion for now to make progress. Maybe we can leave the race be since it's probably quite rare that it would trip someone.

Reply:
Yes, totally agree about potential race between controller and speaker pods.

Probably, I've found solution for the DaemonSet wait for ready condition:
```
kubectl wait -n metallb-system daemonsets.apps speaker --for=jsonpath='{.status.numberReady}'=$(kubectl get daemonsets.apps speaker -n metallb-system  -ojsonpath='{.status.currentNumberScheduled}'
```

Reply:
Thanks, but it doesn't work: When trying to run

```
kubectl wait pods -n metallb-system -l app=metallb,component=speaker --for condition=Ready --timeout 5m
```

right after

```
kubectl wait -n metallb-system daemonsets.apps speaker \
  --for jsonpath='{.status.numberReady}'="$(kubectl get daemonsets.apps speaker -n metallb-system -ojsonpath='{.status.currentNumberScheduled}')" --timeout 5m
```

```
error: no matching resources found
```

Reply:
It's strange, but let's keep the original approach. Thank you.

Reply:
Actually I've already adopted your suggestion partially, just without waiting for the daemonset (see current state of code). Think we should go back to the `sleep` loops?

Reply:
No, adopted part is good. Without waiting for the DaemonSet approach, as it's seems to not working in some cases.

Probably this code should work. I tried to run it locally and it's working for me
```
kubectl wait -n metallb-system daemonsets.apps -l app=metallb,component=speaker --for=jsonpath='{.status.numberReady}'=$(kubectl get daemonsets.apps -l app=metallb,component=speaker -n metallb-system  -ojsonpath='{.items[0].status.currentNumberScheduled}')
```

Reply:
It's "working" for me only because by the time the controller deployment has converged the speaker DS has converged, too. I ran

```
kubectl wait -n metallb-system daemonsets.apps -l app=metallb,component=speaker \
  --for=jsonpath='{.status.numberReady}'="$(kubectl get daemonsets.apps -l app=metallb,component=speaker -n metallb-system  -ojsonpath='{.items[0].status.currentNumberScheduled}')"

kubectl wait pods -n metallb-system -l app=metallb,component=speaker --for condition=Ready --timeout 5m
```

for testing and got:

```
error: no matching resources found
```

Reply:
Let's keep the current state and if anyone ever complains about the race we switch to `sleep`.